service/listenerpipe.go

```diff
@@ -0,0 +1,61 @@
+package service
+
+import (
+	"errors"
+	"net"
+	"sync"
+)
+
+// ListenerPipe returns a full-duplex in-memory connection, like net.Pipe.
+// Unlike net.Pipe one end of the connection is returned as an object
+// satisfying the net.Listener interface.
+// The first call to the Accept method of this object will return a net.Conn
+// connected to the other net.Conn returned by ListenerPipe.
+// Any subsequent calls to Accept will block until the listener is closed.
+func ListenerPipe() (net.Listener, net.Conn) {
+	conn0, conn1 := net.Pipe()
+	return &preconnectedListener{conn: conn0, closech: make(chan struct{})}, conn1
+}
+
+// preconnectedListener satisfies the net.Listener interface by accepting a
+// single pre-established connection.
+// The first call to Accept will return the conn field, any subsequent call
+// will block until the listener is closed.
+type preconnectedListener struct {
+	accepted bool
+	conn     net.Conn
+	closech  chan struct{}
+	closeMu  sync.Mutex
+	acceptMu sync.Mutex
+}
+
+// Accept returns the pre-established connection the first time it's called,
+// it blocks until the listener is closed on every subsequent call.
+func (l *preconnectedListener) Accept() (net.Conn, error) {
+	l.acceptMu.Lock()
+	if !l.accepted {
+		l.accepted = true
+		l.acceptMu.Unlock()
+		return l.conn, nil
+	}
+	l.acceptMu.Unlock()
```

Comment:
I'd say let's just defer the mutex unlock right after the lock. We block in the function anyways so blocking on a mutex instead of a channel read shouldn't make any difference, and will clean up the code a bit.